src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -215,113 +217,80 @@ private Optional<ThingAssociations> getThingsAssociatedWithCoreDevice() {
-            } else {
-                certificateRegistry.deleteCertificate(certificate);
-            }
-        });
+        for (String thingName : thingNamesAttachedToCore) {
+            Set<String> certificateIdAttachedToThing = this.refreshCertificateAttachments(thingName);
+            certificatesAttachedToThings.addAll(certificateIdAttachedToThing);
+        }
+
+        for (String certificateId: certificatesAttachedToThings) {
+            this.refreshCertificateValidity(certificateId);
+        }
+
+        // Clean up the registries by providing the names of the things that should still be attached and
+        // the ids of the certificates that are still attached to a thing.
+        cleanUpRegistries(thingNamesAttachedToCore, certificatesAttachedToThings);
     }
 
     @SuppressWarnings("PMD.AvoidCatchingGenericException")
-    private void refreshCertificateThingAttachment(Thing thing) {
-        VerifyThingAttachedToCertificate useCase = useCases.get(VerifyThingAttachedToCertificate.class);
-
-        thing.getAttachedCertificateIds().keySet().forEach(certificateId -> {
-            Optional<String> certPem = pemStore.getPem(certificateId);
-
-            if (!certPem.isPresent()) {
-                logger.atWarn()
-                        .kv("certificateId", certificateId)
-                        .kv("thing", thing.getThingName())
-                        .log("Tried to refresh certificate thing attachment but its pem was not found");
-                return;
-            }
+    private Set<String> refreshCertificateAttachments(String thingName) {
+        Thing thing = thingRegistry.getThing(thingName);
+
+        if (Objects.isNull(thing)) {
+            logger.atDebug().kv("thingName", thingName)
+                    .log("No local version found for thing. Not refreshing thing certificate attachments");
+            return Collections.emptySet();
+        }
 
+        Set<String> thingCertificateIds = thing.getAttachedCertificateIds().keySet();
+
+        for (String certificateId: thingCertificateIds) {
             try {
-                Certificate certificate = Certificate.fromPem(certPem.get());
-                useCase.apply(new VerifyThingAttachedToCertificateDTO(thing, certificate));
-            } catch (InvalidCertificateException | RuntimeException e) {
-                logger.atWarn().cause(e).kv("thing", thing.getThingName()).kv("certificate", certificateId)
-                        .log("Failed to verify thing attached to certificate");
+                useCases.get(VerifyThingAttachedToCertificate.class).apply(
+                        new VerifyThingAttachedToCertificateDTO(thingName, certificateId));
+            } catch (RuntimeException e) {
+                logger.atWarn().cause(e).kv("thingName", thing.getThingName()).kv("certificate", certificateId)
+                        .log("Failed to verify thing certificate - certificate pem is invalid");
             }
```

Comment:
Do we anticipate RuntimeException here? How do we know it is because of the invalid PEM?

Reply:
The `VerifyThingAttachedToCertificate` doesn't really requires a certificate, we only need the certificateId so I am passing the primitive to the DTO and the useCase can decide what to do. We won't get an `InvalidCertificateException` given we are not building a certificate. The reason I am catching a generic `RuntimeException` is due to the fact that if someone makes a mistake on the `VerifyThingAttachedToCertificate` and introduces one, then the scheduled thread will die and the refresh would stop.

Right now, the way a runtime exception can happen is if the cloud throws an internal server error or a throttle error and without this we would be on the scenario I described above.

Reply:
To vaibhav's point, if RuntimeException happens on ISE or throttle, then we should probably leave out "certificate pem is invalid"

Also, might be worth having a catch-all in the `run` method to provide total assurance the thread won't get killed

Reply:
+1 to this, we should do a complete audit of all logs. Can handle that separately.

Also, our exception handling story is not quite good enough for me to feel comfortable with these runtime exceptions. Let's also take a closer look at that. Ideally use cases won't throw them